src/java.base/share/classes/jdk/internal/classfile/impl/SplitConstantPool.java

```diff
@@ -0,0 +1,613 @@
+
+    @Override
+    public void writeTo(BufWriter buf) {
+        int writeFrom = 1;
+        buf.writeU2(entryCount());
+        if (parent != null && buf.constantPool().canWriteDirect(this)) {
+            parent.writeConstantPoolEntries(buf);
+            writeFrom = parent.entryCount();
+        }
+        for (int i = writeFrom; i < entryCount(); ) {
+            PoolEntry info = entryByIndex(i);
+            info.writeTo(buf);
+            i += info.poolEntries();
+        }
+    }
+
+    private EntryMap<PoolEntry> map() {
+        if (map == null) {
+            map = new EntryMap<>(Math.max(size, 1024), .75f) {
+                @Override
+                protected PoolEntry fetchElement(int index) {
+                    return entryByIndex(index);
+                }
+            };
+            // Doing a full scan here yields fall-off-the-cliff performance results,
+            // especially if we only need a few entries that are already
+            // inflated (such as attribute names).
+            // So we inflate the map with whatever we've got from the parent, and
+            // later, if we miss, we do a one-time full inflation before creating
+            // a new entry.
+            for (int i=1; i<parentSize; i++) {
+                PoolEntry cpi = parent.cp[i];
+                if (cpi != null)
+                    map.put(cpi.hashCode(), cpi.index());
+            }
+            for (int i = Math.max(parentSize, 1); i < size; ) {
+                PoolEntry cpi = myEntries[i - parentSize];
+                map.put(cpi.hashCode(), cpi.index());
+                i += cpi.poolEntries();
+            }
+        }
+        return map;
+    }
+
+    private void fullScan() {
+        for (int i=1; i<parentSize;) {
+            PoolEntry cpi = parent.entryByIndex(i);
+            map.put(cpi.hashCode(), cpi.index());
+            i += cpi.poolEntries();
+        }
+        doneFullScan = true;
+    }
+
+    private EntryMap<ConcreteBootstrapMethodEntry> bsmMap() {
+        if (bsmMap == null) {
+            bsmMap = new EntryMap<>(Math.max(bsmSize, 16), .75f) {
+                @Override
+                protected ConcreteBootstrapMethodEntry fetchElement(int index) {
+                    return bootstrapMethodEntry(index);
+                }
+            };
+            for (int i=0; i<parentBsmSize; i++) {
+                ConcreteBootstrapMethodEntry bsm = parent.bootstrapMethodEntry(i);
+                bsmMap.put(bsm.hash, bsm.index);
+            }
+            for (int i = parentBsmSize; i < bsmSize; ++i) {
+                ConcreteBootstrapMethodEntry bsm = myBsmEntries[i - parentBsmSize];
+                bsmMap.put(bsm.hash, bsm.index);
+            }
+        }
+        return bsmMap;
+    }
+
+    private <E extends PoolEntry> E internalAdd(E cpi) {
+        return internalAdd(cpi, cpi.hashCode());
+    }
+
+    private <E extends PoolEntry> E internalAdd(E cpi, int hash) {
+        int newIndex = size-parentSize;
```

Comment:
I'm not sure I get this. The new entry is constructed with an index set to the pool size. That said, when we build the entry we don't yet know if the entry is already in the pool. The EntryMap data structure seems to have logic to detect duplicates (e.g. adding an already added entry is a no-op) - but this method seems to (a) add the CP entry to the `newEntries` array anyway and (b) happily return the entry with the index set to the pool size (which might, or might not, be the correct final index). What am I missing?

Reply:
Nevermind - this method is only called if the entry is not found (using `findEntry`) :-)